### PR TITLE
#3332 add devcontainer

### PR DESCRIPTION
**Pull request by paule96**
This PR adds the first try to enable DevContainers for the Aspire project.
Currently, the functionality in the codespaces of GitHub is very limited. It's enough to write tests and debug them. But to run the playground examples it's not enough because something with the dashboard is wrong there.
For DevContainers on a local installation, it works quite well, if you can provide enough memory.

Currently what takes quite long when you first start the DevContainer are two things:

- creating all the base layers that will add docker / powershell / dapr
- the restore of all the projects in the repository after the devcontainer is up and running
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3520)

### CODE REVIEW

File: docs/machine-requirements.md [hunk 1]
```diff
@@ -2,20 +2,51 @@
...
+* [Windows](https://docs.docker.com/desktop/install/windows-install/)
+* [MacOS X](https://docs.docker.com/desktop/install/mac-install/)
+* [Linux](https://docs.docker.com/desktop/install/linux-install/)
+
+## (Browser) Codespaces
+
+Just start the Codespaces in your fork. The initialisation of the code space takes around 5 mins. After that you can open the solution.
+This will take on the free version of Codesapce around 10 mins.
+
+> Warning: With the free version of Codespaces the development expiriance is not nice. We recommend to use at least a code space with 16GB of RAM or use your local VsCode / DevContainers instance.
```
Comment (davidfowl) [inline docs/machine-requirements.md:50]:
```suggestion
> Warning: With the free version of Codespaces the development experience is not nice. We recommend using at least a code space with 16GB of RAM or use your local VSCode / DevContainers instance.
```

**Comment by paule96** [comment]
There is also the issue, that for the playground things to work, you need to set in the `launch.json` the `ASPNETCORE_URLS`. I don't know if this is currently a bug in all playground samples or if this should be working.

File: docs/machine-requirements.md (hunk 1 above)
Comment (paule96) [reply docs/machine-requirements.md]: fixed

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		"ghcr.io/devcontainers/features/powershell:1": {},
+		"ghcr.io/dapr/cli/dapr-cli": {
+			"version": "1.12.0"
+		}
+	},
+	"customizations": {
+		"vscode": {
+			"extensions": [
+				"ms-dotnettools.vscodeintellicode-csharp",
+				"chneau.resource-monitor"
```
Comment (danmoseley) [inline .devcontainer/devcontainer.json:18]: I'm not familiar with this. Is it the most mainstream option? Doesn't seem super popular.

https://marketplace.visualstudio.com/items?itemName=chneau.resource-monitor

Comment (paule96) [reply .devcontainer/devcontainer.json:18]: oh right sorry. It's a personal preference to see if the dev container hangs because it hangs or if it really does something. I will remove it

**Comment by danmoseley** [comment]
@eerhardt you did the codespaces for dotnet/runtime. Does this look reasonable?

**Comment by danmoseley** [comment]
As far as prebuild, I think (?) after this goes in I just go switch it on?

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/dotnet
+{
+	"name": "C# (.NET)",
```
Comment (danmoseley) [inline .devcontainer/devcontainer.json:4]: I think the standard in this repo is 4 spaces. can you please remove tabs?

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,44 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/dotnet
```
Comment (danmoseley) [inline .devcontainer/devcontainer.json:2]: I haven't read the devcontainer spec. The aspnetcore one has a `build {..}` section, this does not?
https://github.com/dotnet/aspnetcore/blob/4eafb8ecdcf25b054b4bcda69ab1ecb498aece10/.devcontainer/devcontainer.json#L5

File: docs/machine-requirements.md [hunk 5]
```diff
@@ -2,20 +2,51 @@
...
 [.NET 8 SDK](https://github.com/dotnet/installer#installers-and-binaries)
 
-## Install Docker Desktop
+### Install Docker Desktop
+
 * [Windows](https://docs.docker.com/desktop/install/windows-install/)
 * [MacOS X](https://docs.docker.com/desktop/install/mac-install/)
 * [Linux](https://docs.docker.com/desktop/install/linux-install/)
+
+## (Windows / Linux / Mac) DevContainer in VSCode
```
Comment (danmoseley) [inline docs/machine-requirements.md:27]:
```suggestion
## (Windows / Linux / Mac) DevContainer in VS Code
```
etc

File: docs/machine-requirements.md [hunk 6]
```diff
@@ -2,20 +2,51 @@
...
+### Install Docker Desktop
+
+* [Windows](https://docs.docker.com/desktop/install/windows-install/)
+* [MacOS X](https://docs.docker.com/desktop/install/mac-install/)
+* [Linux](https://docs.docker.com/desktop/install/linux-install/)
+
+## (Browser) Codespaces
+
+Just start the Codespaces in your fork. The initialisation of the code space takes around 5 mins. After that you can open the solution.
+This will take on the free version of Codesapce around 10 mins.
```
Comment (danmoseley) [inline docs/machine-requirements.md:48]:
```suggestion
This will take on the free version of Codespace around 10 mins.
```

File: docs/machine-requirements.md [hunk 7]
```diff
@@ -2,20 +2,51 @@
...
+
+> Warning: With the free version of Codespaces the development experience is not nice. We recommend using at least a Codespace with 16GB of RAM or use your local VSCode / DevContainers instance.
```
Comment (danmoseley) [inline docs/machine-requirements.md:50]: I think there is a setting somewhere that I can use to increase this (billed to the .NET foundation..)

File: .gitignore [hunk 8]
```diff
@@ -144,3 +144,6 @@ node_modules/
 /playground/**/.gitignore
 /playground/**/azure.yaml
 /playground/**/next-steps.md
+
+# Ignore mono files on linux
+.mono/
```
Comment (danmoseley) [inline .gitignore:149]: where do these come from?

**Comment by kvenkatrajan** [comment]
Overall this looks good. To allow for debugging of playground apps additional considerations would need to be applied : 

- version of C# dev kit > 1.5.10 
Its currently in pre-release 
image
This version has some changes to debug/run in C# dev kit that would help the dev inner loop for aspire apps and we are yet to push the changes to release
- currently while running the TestShop app in codespaces : _System.AggregateException: One or more errors occurred. (Unable to configure HTTPS endpoint. No server certificate was specified, and the default developer certificate could not be found or is out of date._ Since current launch behavior is defaulted to https  FYI : ALLOW_UNSECURED_TRANSPORT. Here is some documentation on how to configure : https://hub.docker.com/_/microsoft-devcontainers-dotnet
- To default to http,  environment variable : ALLOW_UNSECURED_TRANSPORT (not recommended) would need to be set.

Rest looks good to me.

CC: @baronfel for additional review.

File: .devcontainer/devcontainer.json [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+		"ghcr.io/devcontainers/features/docker-in-docker:2": {},
+		"ghcr.io/devcontainers/features/powershell:1": {},
+		"ghcr.io/dapr/cli/dapr-cli": {
+			"version": "1.12.0"
+		}
+	},
+	"customizations": {
+		"vscode": {
+			"extensions": [
+				"ms-dotnettools.vscodeintellicode-csharp"
```
Comment (baronfel) [inline .devcontainer/devcontainer.json:17]: why this extension instead of just devkit?

File: .devcontainer/devcontainer.json [hunk 10]
```diff
@@ -0,0 +1,44 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/dotnet
+{
+	"name": "C# (.NET)",
+	// Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+	"image": "mcr.microsoft.com/devcontainers/dotnet:1-8.0-bookworm",
+	"features": {
+		"ghcr.io/devcontainers/features/docker-in-docker:2": {},
+		"ghcr.io/devcontainers/features/powershell:1": {},
+		"ghcr.io/dapr/cli/dapr-cli": {
```
Comment (eerhardt) [inline .devcontainer/devcontainer.json:10]: Why do we need the dapr-cli? It should only be needed if someone is actually working on the dapr extension/playground app, right?

File: docs/machine-requirements.md (hunk 7 above)
Comment (paule96) [reply docs/machine-requirements.md:50]: I don't know how it works for members of the .NET Foundation or Microsoft employees, but for the rest of the world you need to pay a bill if you want to use it.

That's why I personally use the local dev containers for repositories like that more often.

File: .gitignore (hunk 8 above)
Comment (paule96) [reply .gitignore:149]: if you make run `dotnet build` on linux they apear

File: .devcontainer/devcontainer.json (hunk 10 above)
Comment (paule96) [reply .devcontainer/devcontainer.json]: yes I agree. This is for me a big question mark too. On the other hand, should someone who wants to contribute, think about that stuff, or should it be just there?

I mean that's the nice thing about devcontainers that you can just start developing, because someone else took care of installing all the tools you need. 

So I don't know what todo here.
If we install dapr we should maybe install other stuff too. (azure emulators, other CLIs, etc....)

File: .devcontainer/devcontainer.json (hunk 9 above)
Comment (paule96) [reply .devcontainer/devcontainer.json]: Because it installs the devkit and you get better autocomplete.

**Comment by paule96** [comment]
> Overall this looks good. To allow for debugging of playground apps additional considerations would need to be applied :
> 
> * version of C# dev kit > 1.5.10
>   Its currently in pre-release
>   image
>   This version has some changes to debug/run in C# dev kit that would help the dev inner loop for aspire apps and we are yet to push the changes to release
> * currently while running the TestShop app in codespaces : _System.AggregateException: One or more errors occurred. (Unable to configure HTTPS endpoint. No server certificate was specified, and the default developer certificate could not be found or is out of date._ Since current launch behavior is defaulted to https. Here is some documentation on how to configure : https://hub.docker.com/_/microsoft-devcontainers-dotnet
> * To default to http,  environment variable : ALLOW_UNSECURED_TRANSPORT (not recommended) would need to be set.
> 
> Rest looks good to me.
> 
> CC: @baronfel for additional review.

@kvenkatrajan for your issue with the HTTPS endpoint the you could run `dotnet dev-certs https`. That should generate the missing certificate.
In your browser you will still get a certificate warning, because your local machine is not knowing / trusting the certificate root. But the CLI can export the generated cert. (`dotnet dev-certs https --export-path ~/mydevcert.pfx --password "securePassword"`) This can be then applied as a trusted root on your OS. (depends on your local OS how)

File: .devcontainer/devcontainer.json [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+	},
+	"customizations": {
+		"vscode": {
+			"extensions": [
+				"ms-dotnettools.vscodeintellicode-csharp"
+			]
+		}
+	},
+	"remoteEnv": {
+		"PATH": "${containerEnv:PATH}:/home/vscode/.dapr/bin"
```
[the remark this answers is not in the document]
Comment (paule96) [reply .devcontainer/devcontainer.json]: yes the dapr CLI is found. But the problem is you cant find the components of dapr without the line. But I need to remember why this was important. (this devcontainer.json was written around 3 Weeks ago)

File: docs/machine-requirements.md (hunk 5 above)
Comment (paule96) [reply docs/machine-requirements.md]: done

File: docs/machine-requirements.md (hunk 6 above)
Comment (paule96) [reply docs/machine-requirements.md]: done

File: .gitignore (hunk 8 above)
Comment (paule96) [reply .gitignore:149]: image

it comes from the devkit as it looks like

**Comment by paule96** [comment]
> Overall this looks good. To allow for debugging of playground apps additional considerations would need to be applied :
> 
> * version of C# dev kit > 1.5.10
>   Its currently in pre-release
>   image
>   This version has some changes to debug/run in C# dev kit that would help the dev inner loop for aspire apps and we are yet to push the changes to release

The problem there is, that I currently don't see a way to specify the use of prerelease version of extensions in devcontainers. If you take a look into the[ spec of devcontianers](https://github.com/devcontainers/spec/blob/main/docs/specs/supporting-tools.md#visual-studio-code), the extensions property is a list of extension ids without version numbers.

**Comment by kvenkatrajan** [comment]
> The problem there is, that I currently don't see a way to specify the use of prerelease version of extensions in devcontainers. If you take a look into the[ spec of devcontianers](https://github.com/devcontainers/spec/blob/main/docs/specs/supporting-tools.md#visual-studio-code), the extensions property is a list of extension ids without version numbers.

I think this should be fine but just giving a headsup that debugging might not work for playground apps in codespaces till this version is live. Maybe worth specifying that in the machine-requirements.md

File: .devcontainer/devcontainer.json [hunk 12]
```diff
@@ -0,0 +1,40 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/dotnet
+{
+  "name": "C# (.NET)",
+  // Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+  "image": "mcr.microsoft.com/devcontainers/dotnet:1-8.0-bookworm",
```
Comment (timheuer) [inline .devcontainer/devcontainer.json:6]: We should just use the 8.0 SDK image here.

File: .devcontainer/devcontainer.json [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    "ghcr.io/devcontainers/features/docker-in-docker:2": {},
+    "ghcr.io/devcontainers/features/powershell:1": {},
+    "ghcr.io/dapr/cli/dapr-cli": {
+      "version": "1.12.0"
+    }
+  },
+  "customizations": {
+    "vscode": {
+      "extensions": [
+        "ms-dotnettools.vscodeintellicode-csharp",
```
Comment (timheuer) [inline .devcontainer/devcontainer.json:17]: recommend changing this to `ms-dotnettools.csdevit`

File: .devcontainer/devcontainer.json (hunk 12 above)
Comment (baronfel) [reply .devcontainer/devcontainer.json:6]: The main differences currently between the devcontainers image and the official dotnet SDK image is that 
* NuGet packages will get xmldocs in the devcontainers image (which we want), and
* The devcontainers image includes Powershell Core

I think those are enough of a value-add (especially the XMLDocs) to justify using this base image.

File: .devcontainer/devcontainer.json [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+{
+  "name": "C# (.NET)",
+  // Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+  "image": "mcr.microsoft.com/devcontainers/dotnet:1-8.0-bookworm",
+  "features": {
+    "ghcr.io/devcontainers/features/docker-in-docker:2": {},
+    "ghcr.io/devcontainers/features/powershell:1": {},
+    "ghcr.io/dapr/cli/dapr-cli": {
+      "version": "1.12.0"
+    }
```
Comment (timheuer) [inline .devcontainer/devcontainer.json:12]: For this repo recommend adding `ghcr.io/azure/azure-dev/azd:0` for convenience on deployment validations as well with Azure

File: .devcontainer/devcontainer.json (hunk 12 above)
Comment (timheuer) [reply .devcontainer/devcontainer.json:6]: Educate me -- why doesn't the SDK image get the xmldocs?

RE: Powershell -- easy add-on as a feature (which is in this as well)

No objections really either, just felt like the actual SDK image felt like a better base -- it's what I use exclusively and just add features.

Comment (eerhardt) [reply .devcontainer/devcontainer.json:6]: > Educate me -- why doesn't the SDK image get the xmldocs?

https://github.com/dotnet/dotnet-docker/issues/2790

TLDR is "performance".

Comment (baronfel) [reply .devcontainer/devcontainer.json:6]: Funnily enough, I was just going to [the source](https://github.com/dotnet/dotnet-docker/blob/main/src/sdk/8.0/bookworm-slim/amd64/Dockerfile#L14) and noticed that powershell core is already on the SDK images - so that removes one of the benefits of the devcontainers image. Kinda feels like we're all over the place with these images, thematically :)

Comment (timheuer) [reply .devcontainer/devcontainer.json:6]: Yep, noticed that too -- this devcontainer image looks to take a patch to the powershell also though, which i would expect the powershell feature latest to have also.  Anyhow, thanks for the edu on the xmldocs -- seems odd our <cough> sdk image </cough> doesn't really match the sdk, but at least now I know!

Comment (eerhardt) [reply .devcontainer/devcontainer.json:6]: > seems odd our sdk image doesn't really match the sdk

It was super surprising to me too when I learned this too. I understand the constraints and get it adds a lot of value to the "non-human-interactive" use cases. So it seems when a human is involved, its best to start with the dev container image.

File: .devcontainer/devcontainer.json (hunk 3 above)
Comment (danmoseley) [reply .devcontainer/devcontainer.json]: Maybe it's not, but as far as I know this repo is 4 spaces everywhere - all the core dotnet repos are.

File: .gitignore (hunk 8 above)
Comment (danmoseley) [reply .gitignore:149]: OK fine.

File: .devcontainer/devcontainer.json (hunk 3 above)
Comment (paule96) [reply .devcontainer/devcontainer.json]: I just did check the runtimes repo [file](https://github.com/dotnet/runtime/blob/main/.devcontainer/devcontainer.json) it uses `tabs`. 
I already changed to spaces, but don't know if I should enable the 4 size indent_size. (default is 2)

I would change it like this:
image

But that means that every file will get the 4 space wide indent.

File: .devcontainer/devcontainer.json (hunk 13 above)
Comment (paule96) [reply .devcontainer/devcontainer.json:17]: are there reasons to it? had the same recommendation from @baronfel.
Isn't the intellicode extension just the more advanced one?

File: .devcontainer/devcontainer.json (hunk 3 above)
Comment (danmoseley) [reply .devcontainer/devcontainer.json]: Ah - ok fine to leave consistent with runtime and we can leave settings as they are for now.

File: .devcontainer/devcontainer.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (paule96) [reply .devcontainer/devcontainer.json]: did test it again. Seems to work without. 🤷‍♀️ removed it

File: docs/machine-requirements.md (hunk 7 above)
Comment (paule96) [reply docs/machine-requirements.md]: image
This is how you can configure. because It don't want to invest money I didn't try what will happen if you click create :D

File: .devcontainer/devcontainer.json (hunk 10 above)
Comment (danmoseley) [reply .devcontainer/devcontainer.json]: I suggest just the basics in this PR and we can add more as requested/needed

File: docs/machine-requirements.md (hunk 7 above)
Comment (danmoseley) [reply docs/machine-requirements.md]: If I remember right, we can simply add as dotnet org members folks like you who want to use this. Then the foundation pays.

File: .devcontainer/devcontainer.json (hunk 9 above)
Comment (danmoseley) [reply .devcontainer/devcontainer.json]: @timheuer ?

File: .devcontainer/devcontainer.json (hunk 10 above)
Comment (paule96) [reply .devcontainer/devcontainer.json]: @danmoseley what are the basics for you? Dapr cli and Azure cli? Just Azure cli? None of both? More?

Comment (danmoseley) [reply .devcontainer/devcontainer.json]: Ha, good point, this repo explicitly isn't cloud specific. @eerhardt any opinion?

Comment (eerhardt) [reply .devcontainer/devcontainer.json]: I've never used the dapr cli and I work in the repo every day.

My opinion would be to not include the dapr cli or the Azure cli initially.

If someone wants these, they can customize their own container right? https://docs.github.com/codespaces/setting-your-user-preferences/personalizing-github-codespaces-for-your-account

File: .devcontainer/devcontainer.json [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+  },
+  // Features to add to the dev container. More info: https://containers.dev/features.
+  // "features": {},
+  // Use 'forwardPorts' to make a list of ports inside the container available locally.
+  // "forwardPorts": [5000, 5001],
+  // "portsAttributes": {
+  //		"5001": {
+  //			"protocol": "https"
+  //		}
```
Comment (RussKie) [inline .devcontainer/devcontainer.json:30]: [nit] please change tabs to spaces

File: docs/machine-requirements.md [hunk 16]
```diff
@@ -2,20 +2,51 @@
...
 * [Windows](https://docs.docker.com/desktop/install/windows-install/)
 * [MacOS X](https://docs.docker.com/desktop/install/mac-install/)
 * [Linux](https://docs.docker.com/desktop/install/linux-install/)
+
+## (Windows / Linux / Mac) DevContainer in VS Code
+
+On Windows you could also use VS Code with the DevContainers extension. This requieres that you have installed a container engine installed.
+Currently it's only tested with Docker Desktop.
+
+> Warning: This will use around 16GB of RAM, after you loaded the solution.
```
Comment (RussKie) [inline docs/machine-requirements.md:32]: What about this?
```suggestion
> :warning: This will use around 16GB of RAM, after you loaded the solution.
```

Ditto below

File: docs/machine-requirements.md [hunk 17]
```diff
@@ -2,20 +2,51 @@
...
+* [Linux](https://docs.docker.com/desktop/install/linux-install/)
+
+## (Browser) Codespaces
+
+Just start the Codespaces in your fork. The initialisation of the code space takes around 5 mins. After that you can open the solution.
+This will take on the free version of Codespace around 10 mins.
+
+> Warning: With the free version of Codespaces the development experience is not nice. We recommend using at least a Codespace with 16GB of RAM or use your local VS Code / DevContainers instance.
+
+> Information: Currently the codespaces can not be used to try / debug the aspire dashboard. This is a known issue. If you want to debug it, use Visual Studio or your local VS Code and the DevContainers.
```
Comment (RussKie) [inline docs/machine-requirements.md:52]:
```suggestion
> Information: Currently the codespaces can not be used to try / debug the Aspire dashboard. This is a known issue. If you want to debug it, use Visual Studio or your local VS Code and the DevContainers.
```

Comment (RussKie) [inline docs/machine-requirements.md:52]: 💭 
```suggestion
> :warning: Currently the codespaces can not be used to try / debug the aspire dashboard. This is a known issue. If you want to debug it, use Visual Studio or your local VS Code and the DevContainers.
```

File: .devcontainer/devcontainer.json (hunk 15 above)
Comment (paule96) [reply .devcontainer/devcontainer.json:30]: @RussKie I had the same discussion earlier with @danmoseley [here](https://github.com/dotnet/aspire/pull/3520#discussion_r1560509909). And we agreed on doing it the same as the runtime team is doing. Are there any new information that where discussed internally?

Comment (RussKie) [reply .devcontainer/devcontainer.json:30]: The rest of the file uses spaces, and this part uses tabs. It just looks inconsistent. That's all.

File: .devcontainer/devcontainer.json (hunk 13 above)
Comment (WeihanLi) [reply .devcontainer/devcontainer.json:18]: https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.vscodeintellicode-csharp
image
seemed we should install `ms-dotnettools.csdevkit` to keep it working

Comment (paule96) [reply .devcontainer/devcontainer.json:18]: If you open the extension page in visual studio you see it will install the C# extension by itself:

Image showing the dependencies of the IntelliCode extension including C# extension

Comment (paule96) [reply .devcontainer/devcontainer.json:18]: @WeihanLi It's the same like for the C# extension. That extension will install [`.NET Install Tool`](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.vscode-dotnet-runtime) Extension.

**Comment by davidfowl** [comment]
@danmoseley @timheuer Is there anything missing from this? It would be great to have this merged.

**Comment by paule96** [comment]
ping @danmoseley and @timheuer :)

**Comment by paule96** [comment]
Okay I will close this PR because I saw on twitter that someone else added dev container support. :/ A bit sad about the not existing communication at the end...

#6491

**Comment by davidfowl** [comment]
mea culpa @paule96, we should have integrated this PR! There were other changes that included making this aspire work in codespaces that were done as well but that's not an excuse.

File: docs/machine-requirements.md [hunk 18]
```diff
@@ -26,3 +30,24 @@ When you install, ensure that both:
 * [Windows](https://podman.io/docs/installation#windows)
 * [macOS](https://podman.io/docs/installation#macos)
 * [Linux](https://podman.io/docs/installation#installing-on-linux)
+
+## (Windows / Linux / Mac) DevContainer in VS Code
```
Comment (davidfowl) [inline docs/machine-requirements.md:34]: @paule96 This would still be very useful to merge

cc @mitchdenny

**Comment by paule96** [comment]
@davidfowl yes but probably a new pullrequest would be better, because merging this now is I guess a bit wrong. Let me see if I find the time today to reopen a new PR with just the changes that are left. My quick analysis is:

- migrate the docs
- add the dapr init post command
- maybe change the base image to the Ubuntu version (has only benefits on windows as far as I know, in combination with the SshD service)
- add the intellicode extension
- add the editorconfig extension

**Comment by mitchdenny** [comment]
@paule96 sorry I didn't see your PR here when I merged mine. I did the work on improving the codespaces support so that the dashboard would show the port forwarded URLs. If you do put up another PR with these changes feel free to tag me in it so we can do a quick review.
